File: 07/Java/TripleExpression/CheckedAbs.java
package TripleExpression;

public class CheckedAbs implements TripleExpression {
	private TripleExpression value;
    public CheckedAbs(TripleExpression val) {
        value = val;
    }

    protected void check(int a) throws Exception {
        if (a == Integer.MIN_VALUE) {
            throw new OverflowException();
        }
    }

    public int evaluate(int x, int y, int z) throws Exception {
        int val = value.evaluate(x, y, z);
        check(val);
        if(val < 0) {
            val = -val;
        }
        return val;
    }
}

File: 07/CSharp/TripleExpressionApp/TripleExpressionApp/CheckedAbs.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    public class CheckedAbs : ITripleExpression
    {
        private ITripleExpression value;
        public CheckedAbs(ITripleExpression value)
        {
            this.value = value;
        }

        protected void Check(double a)
        {
            if (a == double.MinValue)
            {
                throw new OverflowException();
            }
        }

        public double Evaluate(double x, double y, double z)
        {
            double val = value.Evaluate(x, y, z);
            Check(val);
            if (val < 0)
            {
                val = -val;
            }
            return val;
        }
    }
}
